### ISSUE TITLE
Overview of Hosts using ServiceSet

### ISSUE THREAD

**Issue by mwaldmueller** [Icinga/icingaweb2-module-director#1674]
## Expected Behavior
Tab "Hosts" at a ServiceSet should show "Hosts using this set".

## Current Behavior
<img width="560" alt="bildschirmfoto 2018-10-10 um 14 25 28" src="https://user-images.githubusercontent.com/3973350/46736113-72ec8a80-cc98-11e8-8a22-18f182fb9f2c.png">

<img width="560" alt="bildschirmfoto 2018-10-10 um 14 25 15" src="https://user-images.githubusercontent.com/3973350/46736146-8c8dd200-cc98-11e8-8d9a-ec54c209a6cf.png">

<img width="554" alt="bildschirmfoto 2018-10-10 um 14 25 04" src="https://user-images.githubusercontent.com/3973350/46736103-6831f580-cc98-11e8-8d8f-72db51673bc8.png">

## Your Environment

* Director version (System - About): V1.5.1
* Icinga Web 2 version and modules (System - About): 2.6.1
* Icinga 2 version (`icinga2 --version`): 2.9.2
* Operating System and version: CentOS 7.5
* Webserver, PHP versions: Apache 2.4, PHP 7.1.8

**Comment by gphargreaves** [comment]
@mwaldmueller Are  you referring to the tab text itself that should be updated?

**Comment by mwaldmueller** [comment]
@gphargreaves No, I would expect a listing of the Hosts that use this ServiceSet.

**Comment by gphargreaves** [comment]
Just having a quick look at this, looks like the query in the back-end will return directly assigned service sets (IE service-sets manually assigned from the host page) however won't process those hosts that 
match the Assign Filter.

Not sure we can combine the results of the query with the results of the assign filter matches (and we probably shouldn't either for usability) 

@Thomas-Gelf not sure if this is intentional? Can you advise?

**Comment by gphargreaves** [comment]
Host Groups also seem to have a similar issue, only direct assignments are displayed and not hosts targeted by the assignment filter.

I will raise a separate issue for this shortly.

I am currently working on a fix for the Service Set issue and will submit the PR for review as soon as possible.

**Comment by gphargreaves** [comment]
@Thomas-Gelf I have a fix for this issue however adding multiple tables to the host tab of the serviceset breaks the quicksearch. My only option at the moment is to disable the quicksearch for this page however I would appreciate your guidance on if this is suitable?

Alternatively is there a way for two quicksearch bars to co-exist on the same view?

**Comment by gphargreaves** [comment]
Followed the same model as the Hosts Service tab and disabled quick search.

Fix submitted in PR #1680

**Comment by gphargreaves** [comment]
After further review of the existing sources I have identified a better way of implementing this. PR #1680 has been closed pending a new PR that implements this in a better way.

This will also resolve the quick search issue with this fix

**Comment by Thomas-Gelf** [comment]
@gphargreaves: sorry for the long silence, I'm pretty busy with other tasks these weeks. As a Service Set can also be an Apply Rule, for a correct implementation also such rules need to be implemented. HostApplyMatches is able to do so, but iterating over thousands of Hosts is expensive. We're usually able to do so in a (very) few seconds be initializing in our ugly PrefetchCache, but that's still slightly too slow for a good UI experience in largish environments.

You already discovered the GroupMembershipResolver, and yes, something similar will be the answer. To avoid redundancy my plan was to first implement #573, this means storing all objects in a flat representation for fast searches. Changing a Template would then recalculate properties for inheriting objects. This will then allow us to delegate filtering to the database.

Parts of this have already been pushed to master by accident long time ago, that's why there is for example the unused table icinga_flat_var in our schema. We should then add tables caching object matches for filter strings per Host and Service. Knowing which properties changed for Hosts or Services will help us to understand which filters should be candidates for invalidation. And this should then trigger re-calculation for those filters.

So, that's the plan so far. Once we have all this, we'll be able to add many more fancy features, and fixing this issue will be a trivial joke. Still, if you want to fix this issue without waiting for these dependencies I'd love to merge your upcoming pull request. If you face specific issue let me know. And ping me twice in case you get no answer ;-)

Thanks for your contributions,
Thomas
